using_restricted_array.rb

```diff
@@ -7,41 +7,94 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  indexes = 0
+  until array[indexes] == nil
+    indexes += 1
+  end
+  indexes
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  new_array = []
+  length = length(array)
+  index = 0
+  until index = length - 1
+    new_array << array[index]
+    puts array[index]
+    index += 1
+
+  end
+
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  (0 .. (length - 1)).each do |index|
+    if array[length] == value_to_find
+      return true
+      exit
+    end
+  end
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  largest = 0
+  (0 .. (length - 1)).each do |index|
+    if array[index] > largest
+      largest = array[index]
+    end
+  end
+  largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  smallest = 10000
+  (0 .. (length - 1)).each do |index|
+    if array[index] < smallest
+      smallest = array[index]
+    end
+  end
+  smallest
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  index = length - 1
```

Comment:
The problem statements asks to reverse the array in place i.e. without using a new array. Please see the slide deck we used in class to see the algorithm for this. Please slack me if you have questions.

This code creates a new array (and the tests shouldn't be passing).